### PR TITLE
Upgrade dependencies

### PR DESCRIPTION
**Pull request by BrendanAnnable**
Significant upgrades with:

* Webpack 5
* Storybook 6 Alpha
* TypeScript 4.3
* Socket.IO 4
* React 17
* Compilation target of ESNext

I believe everything works, except I had to remove volatile emits due to this current SocketIO bug: https://github.com/socketio/socket.io/issues/3919 — we might have to block this PR on that 😕 

Things I tested:

`yarn dev`
`yarn build`
`yarn storybook`
`yarn storybook:build`
`yarn test`

### CODE REVIEW

File: nusight2/src/client/components/dashboard/model.ts [hunk 1]
```diff
@@ -7,7 +7,7 @@ import { DashboardRobotModel } from './dashboard_robot/model'
 import { FieldModel } from './field/model'
 
 export class DashboardModel {
-  @observable private robotModels: RobotModel[]
+  @observable.ref private robotModels: RobotModel[]
```
[the remark this answers is not in the document]
Comment (BrendanAnnable) [reply nusight2/src/client/components/dashboard/model.ts:10]:
```suggestion
  @observable private robotModels: RobotModel[]
```

File: nusight2/tsconfig.json [hunk 2]
```diff
@@ -3,7 +3,8 @@
     "allowJs": true,
     "esModuleInterop": true,
     "sourceMap": true,
-    "target": "ES2019",
+    "target": "ESNext",
+    "useDefineForClassFields": false,
```
Comment (BrendanAnnable) [inline nusight2/tsconfig.json:7]: mobx completely breaks without this

File: nusight2/webpack.config.ts [hunk 3]
```diff
@@ -62,34 +62,28 @@ export function getClientConfig({
...
+            {
+              loader: 'postcss-loader',
+              options: {
+                postcssOptions: {
                   plugins: [
-                    require('postcss-import')({ addDependencyTo: webpack }),
-                    require('postcss-url')(),
-                    require('postcss-cssnext')(),
-                    require('postcss-reporter')(),
-                    require('postcss-browser-reporter')({ disabled: isProduction }),
```
Comment (BrendanAnnable) [inline nusight2/webpack.config.ts:87]: removed, I don't think we use these

File: nusight2/webpack.config.ts [hunk 4]
```diff
@@ -124,7 +118,6 @@ export function getClientConfig({
     },
     optimization: {
       splitChunks: {
-        name: true,
```
Comment (BrendanAnnable) [inline nusight2/webpack.config.ts:127]: not needed anymore IIUC

File: nusight2/webpack.config.ts [hunk 5]
```diff
@@ -185,7 +170,7 @@ export function getServerConfig({
     resolve: {
       extensions: ['.js', '.ts'],
     },
-    externals: [nodeExternals()],
+    externals: [nodeExternals()] as webpack.Configuration['externals'],
```
Comment (BrendanAnnable) [inline nusight2/webpack.config.ts:173]: seems to be a webpack 4 vs 5 typing problem

**Review by TrentHouliston — APPROVED** [review]
The parts that I understand are fine, the rest doesn't seem wrong :P

**Comment by BrendanAnnable** [comment]
I'll take it! (if anyone else sees a problem though, just let me know)

**Comment by JosephusPaye** [comment]
Looks like the socket.io issue with volatile sending of binary data is fixed as of https://github.com/socketio/socket.io/commit/dc81fcf461cfdbb5b34b1a5a96b84373754047d5, but it's not released yet.

**Pull request merged**